pilot/pkg/model/context.go

```diff
@@ -362,6 +368,16 @@ func ParseServiceNodeWithMetadata(s string, metadata map[string]string) (*Proxy,
 	out.ID = parts[2]
 	out.DNSDomain = parts[3]
 	out.IstioVersion = ParseIstioVersion(metadata[NodeMetadataIstioVersion])
+
+	if data, ok := metadata[NodeMetadataLabels]; ok {
+		nodeLabels := make(map[string]string)
+		if err := json.Unmarshal([]byte(data), nodeLabels); err != nil {
+			return out, fmt.Errorf("invalid %s: %s", NodeMetadataLabels, data)
```

Comment:
shouldn't we fallback to old behavior? does node metadata always have the workload labels?

Reply:
This shouldn't happen, as we already have 2 tries.

Reply:
What do you mean two tries? If I connect to Pilot without the "proper" node metadata then this will return an error which will cause the whole discovery request to fail

Reply:
node metadata starting with 1.3 *and* using standard means of injection *will* have labels. However 1.3 pilot should be able to service 1.2 proxy so anything that assumes labels here will not work for all cases.

Reply:
Is "standard means of injection" a requirement for istio? What about VMs? And like you mentioned, 1.2 compatibility is important

Reply:
Standard means of injection will never be a requirement. We *can* require or recommend that "labels be present in bootstrap", but we have not done so. Therefore it is important to have fallback in the code.

Reply:
This code wont be hit mostly. Older 1.2 proxies wont have the node metadata with LABELS field. So they will automatically fall back to getting that info from env.GetProxy... For newer proxies, if one supplies the LABELS field in the node metadata, then it better be correct kv pairs. Arbitrary junk in there should be rejected as he is doing currently.